tests/test_smbclient_shutil.py

```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2019, Jordan Borean (@jborean93) <jborean93@gmail.com>
+# MIT License (see LICENSE or https://opensource.org/licenses/MIT)
+
+from __future__ import division
+
+import os
+import sys
+
+import pytest
+import re
+from time import sleep
+
+from smbclient import open_file, mkdir, stat, symlink
+from smbclient.path import exists
+from smbclient.shutil import rmtree, copy, copytree_copy
+
+
+def test_rmtree(smb_share):
+    mkdir("%s\\dir2" % smb_share)
+    mkdir("%s\\dir2\\dir3" % smb_share)
+
+    with open_file("%s\\dir2\\dir3\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    with open_file("%s\\dir2\\file2" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    if os.name == "nt" or os.environ.get('SMB_FORCE', False):
+        symlink("%s\\dir2\\file2" % smb_share, "%s\\dir2\\file3" % smb_share)
+
+    assert exists("%s\\dir2" % smb_share) is True
+
+    rmtree("%s\\dir2" % smb_share)
+
+    assert exists("%s\\dir2" % smb_share) is False
+
+
+def test_rmtree_non_existing(smb_share):
+    expected = "[NtStatus 0xc0000034] No such file or directory: "
+    with pytest.raises(OSError, match=re.escape(expected)):
+        rmtree("%s\\dir2" % smb_share)
+
+
+def test_rmtree_non_existing_with_error_callback(smb_share):
+    callback_args = []
+
+    def callback(*args):
+        callback_args.append(args)
+
+    assert not callback_args
+
+    rmtree("%s\\dir2" % smb_share, onerror=callback)
+
+    assert callback_args
+
+
+def test_rmtree_non_existing_ignore_errors(smb_share):
+    rmtree("%s\\dir2" % smb_share, ignore_errors=True)
+
+
+def test_copy(smb_share):
+    mkdir("%s\\dir2" % smb_share)
+    with open_file("%s\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.1)
+    copy("%s\\file1" % smb_share, "%s\\dir2\\file1" % smb_share)
+
+    src_stat = stat("%s\\file1" % smb_share)
+    dst_stat = stat("%s\\dir2\\file1" % smb_share)
+
+    assert src_stat.st_size == dst_stat.st_size
+
+
+def test_copy_raises_when_source_and_target_identical(smb_share):
+    mkdir("%s\\dir2" % smb_share)
+    with open_file("%s\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.1)
+    if (sys.version_info > (3, 0)):
```

Comment:
Why the version difference?

Reply:
That's a pytest bug. On python2, pytest cannot properly handle unicode error messages. I add a comment with a reference to the bug.

Reply:
That shouldn’t be an issue, the `smbprotocol._text` has a `to_native` function which can be used on Python 2 and 3 to create a native string from a Unicode string. Technically it’s a no-op on py3 but on py2 it should overcome the limitation of non-ASCII chars in an exception string.

You should just be able to do `match=re.escape(to_native(expected))` but best to test it out first.

Reply:
With that change, I obtain the same error as before:
```
        expected = 'are the same file'
        with pytest.raises(ValueError, match=re.escape(to_native(expected))):
>           copy("%s\\file1" % smb_share, "%s\\file1" % smb_share)
E           UnicodeEncodeError: 'ascii' codec can't encode character u'\xfd' in position 30: ordinal not in range(128)
```

Reply:
I was not able to resolve this for now. More ideas? I think the corresponding issue is: https://github.com/pytest-dev/pytest/issues/5478

Reply:
Make sure you do to_native when you raise the actual exception itself. Also it’s best to make sure your expectation is a Unicode string first by putting the u prefix before it.

If that doesn’t work we can leave it for now.

Reply:
I tried
```
    if _samefile(src, dst, **kwargs):
        raise ValueError(to_native(u"`%s` and `%s` are the same file") % (src, dst))
```
Same error.

Reply:
No worries, I'll have a look at it later and will just keep it there if I can't fix it. Thanks for trying though.